drivers/scheduler/k8s/k8s.go

```diff
@@ -1801,6 +1801,69 @@ func (k *k8s) createMigrationObjects(
 	return nil, nil
 }
 
+func (k *k8s) getScName(pvc *v1.PersistentVolumeClaim) string {
+	if pvc != nil && pvc.Labels != nil {
+		for key, value := range pvc.Labels {
+			if key == "volume.beta.kubernetes.io/storage-class" {
+				return value
+			}
+		}
+	}
+	return ""
+}
+
+func (k *k8s) getPodsUsingStorage(pods []v1.Pod, provisioner string) []v1.Pod {
+	k8sOps := k8s_ops.Instance()
+	podsUsingStorage := make([]v1.Pod, 0)
+	for _, pod := range pods {
+		for _, volume := range pod.Spec.Volumes {
+			if volume.PersistentVolumeClaim == nil {
+				continue
+			}
+			pvc, err := k8sOps.GetPersistentVolumeClaim(volume.PersistentVolumeClaim.ClaimName, pod.Namespace)
+			if err != nil {
+				logrus.Errorf("failed to get pvc [%s] %s. Cause: %v", volume.PersistentVolumeClaim.ClaimName, pod.Namespace, err)
+				return podsUsingStorage
+			}
+			scName := ""
+			if pvc.Spec.StorageClassName != nil {
+				scName = *pvc.Spec.StorageClassName
+			} else {
+				scName = k.getScName(pvc)
+			}
+			sc, err := k8sOps.GetStorageClass(scName)
+			if err != nil {
+				logrus.Errorf("failed to retrieve storageclass %s. Cause: %v", *pvc.Spec.StorageClassName, err)
+				return podsUsingStorage
+			}
+			if sc.Provisioner == provisioners[provisioner] {
+				podsUsingStorage = append(podsUsingStorage, pod)
+				break
+			}
+		}
+	}
+	return podsUsingStorage
+}
+
+func (k *k8s) PrepareNodeToDecommission(n node.Node, provisioner string) error {
+	k8sOps := k8s_ops.Instance()
+	pods, err := k8sOps.GetPodsByNode(n.Name, "")
+	if err != nil {
+		return &scheduler.ErrFailedToDecommissionNode{
+			Node:  n,
+			Cause: fmt.Sprintf("Failed to get pods on the node: %v. Err: %v", n.Name, err),
+		}
+	}
+	podsUsingStorage := k.getPodsUsingStorage(pods.Items, provisioner)
+	if err = k8sOps.DrainPodsFromNode(n.Name, podsUsingStorage, defaultTimeout, defaultRetryInterval); err != nil {
```

Comment:
timeout for drain should be proportional to the number of pods. For every 40 pods, double the timeout.